proposals/p6333.md

```diff
@@ -0,0 +1,586 @@
+[Clang modules](https://clang.llvm.org/docs/Modules.html) might be used for
+improving Clang compile performance without significant support from Carbon.
+
+For now we will rely on whatever caching Bazel does for the `.a` output of a
+`carbon_library`. No other outputs will be made available. That may change, but
+leads don't want support to be built right now.
+
+## Rationale
+
+-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem)
+    -   `carbon build` should support easy experimentation with Carbon, and also
+        small projects.
+    -   Other build support is intended to scale up for larger codebases.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   The intent is to be able to migrate from a CMake, Makefile, or other
+        build at relatively low cost. An invocation to `clang` can typically be
+        replaced with `carbon clang`, linking a binary becomes `carbon link`,
+        and so on.
+    -   Similarly, `carbon_library` and `carbon_binary` are important to us for
+        Bazel support and a migration from `cc_library` and `cc_binary`.
+
+## Alternatives considered
+
+### Naming of commands and rules
+
+For `carbon compile` and `carbon build`, this is trying to split apart concepts.
+Some considered alternatives are:
+
+-   Merge `compile`, and possibly also `link`, into `build`. Flags could be used
+    to differentiate between the versions desired, rather than subcommand names.
+    -   We expect that splitting these apart makes it easier to turn them into
+        replacements in C++ builds, and easier to understand even in
+        Carbon-specific builds.
+-   Some other name for `public_deps`
+    -   I'm not really sure `public_deps` is the best name, but it's my best
+        thought for something to reflect the transitive nature of API exposure
+        being proposed.
+-   Have `carbon build` produce `a.out`
+    -   `a.out` is the default output of most C++ compilers, but it reflects a
+        legacy executable file format. Using the legacy name may reflect
+        backwards compatibility that Carbon doesn't plan.
+    -   Changing the default output name is probably low-cost, and people will
+        get used to it.
+
+### apis versus hdrs
+
+For `apis` in specific, there are a couple nuanced naming choices due to
+`cc_library`'s `hdrs` naming. This is inherently a divergence, and we could have
+continued using `hdrs`.
+
+Advantages:
+
+-   Naming consistency with C++.
+-   Naming it `apis` could lead to confusion that it should contain _all_ API
+    files, not just ones intended to be public; however, similar confusion could
+    exist with `hdrs`.
+
+Disadvantages:
+
+-   `apis` more clearly reflects the design intent of "API". `hdrs` is an opaque
+    naming choice for Carbon when considered independently of C++.
```

Comment:
Also, `hdrs` are generally not separately compiled to produce `.o` files, whereas `apis` are.